### PR TITLE
gui for race game

### PR DESCRIPTION
**Pull request by tovernaar123**
adding a gui for the race game and a way to add for future games

### CODE REVIEW

File: config/expcore/roles.lua [hunk 1]
```diff
@@ -206,6 +207,7 @@ Roles.define_role_order{
 
 Roles.override_player_roles{
     ["Cooldude2606"]={"Senior Administrator","Moderator","Senior Backer","Supporter"},
+    ["tovernaar123"]={"Senior Administrator"},
```
Comment (Cooldude2606) [inline config/expcore/roles.lua:210]: Would be best not to commit a change used for debuging.
I would imagine that at some point we will replace this list to include all the organisers.

File: expcore/Mini_games.lua [hunk 2]
```diff
@@ -77,6 +74,14 @@ end
 function Mini_games._prototype:add_stop_function(stop_function)
     self.stop_function = stop_function   
 end
+function Mini_games._prototype:add_gui_element(gui_element)
+    self.gui = gui_element   
+end
+
+function Mini_games._prototype:add_gui_callback(callback)
+    self.gui_callback = callback   
+end
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:83]: I know that you have used `add` for your other functions; but, would it not make more sense to call them `set` after all your are setting a value rather than adding to a list. Maybe you want to change this for all functions that are like this?

File: expcore/Mini_games.lua [hunk 3]
```diff
@@ -96,16 +101,19 @@ function Mini_games._prototype:add_event(event_name,func)
...
-        return "Wrong number of arguments"
+    if parse_args then
+        if  mini_game.options ~= #parse_args then
+            return "Wrong number of arguments"
+        end
+    else 
+        if mini_game.options ~= 0 then
+            return "Wrong number of arguments"
+        end
     end
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:117]: Why do you return a string error message? Should you not be using error or assert to check and raise an error when you get unexpected arguments.

File: expcore/Mini_games.lua [hunk 4]
```diff
@@ -189,6 +197,10 @@ function Mini_games.stop_game()
     for i,command_name  in ipairs(mini_game.commands) do
         Commands.disable(command_name)
     end
+    for i,player in ipairs(game.connected_players) do
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:200]: I think game.players and game.connected_players require you to use pairs but I could be wrong about that.

File: expcore/Mini_games.lua [hunk 5]
```diff
@@ -199,75 +211,77 @@ function Mini_games.error_in_game(error_game)
...
+    --main_gui[1] = element.parent.parent.parent.parent.parent.parent
+    for i,player in ipairs(game.connected_players) do
+        main_gui[i] = Gui.get_left_element(player,mini_game_list)
+        Gui.toggle_left_element(player,main_gui[1],false)
+    end
+    if args then
+        Mini_games.start_game(name,args)
+    else 
+        Mini_games.start_game(name)
+    end
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:232]: Are you testing for nil here? If so then you dont need to test since if it is nil then you would be passing nil as the second argument in both caes, making the two calles to start_game identical.

File: expcore/Mini_games.lua [hunk 6]
```diff
@@ -199,75 +211,77 @@ function Mini_games.error_in_game(error_game)
...
 end
 
 
+
 local vote_button =
 Gui.element{
     type = 'sprite-button',
     sprite = 'utility/check_mark',
-    style = 'quick_bar_slot_button',   
+    style = 'quick_bar_slot_button',
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:242]: With new gui update this should either be `slot_button` or `slot_sized_button`

File: expcore/Mini_games.lua [hunk 7]
```diff
@@ -199,75 +211,77 @@ function Mini_games.error_in_game(error_game)
...
-end)
-
 local add_mini_game =
-Gui.element(function(_,parent)
-    local vote_flow = parent.add{ type = 'flow', }
+Gui.element(function(_,parent,name)
+    local vote_flow = parent.add{ type = 'flow', name = name }
     vote_flow.style.padding = 0
-    vote_button(vote_flow)
+    vote_button(vote_flow,name)
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:251]: If vote button is used `Gui.element{}` syntax then why are you passing name? All arguments after the first are ignored unless you are using a custom draw function or a custom style function.

File: config/expcore/roles.lua (hunk 1 above)
Comment (tovernaar123) [reply config/expcore/roles.lua:210]: this is already on master but ok XD

File: expcore/Mini_games.lua (hunk 2 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:83]: yeah i gues that does make more sence

File: expcore/Mini_games.lua (hunk 3 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:117]: i return the string for when /start is used as this will print it out as an error

File: expcore/Mini_games.lua (hunk 4 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:200]: I am  sure your wrong here, as i tested it it does not use the event.player_index but it is an array.  game.connected_players["name"] is not even valid i also tested that.

File: expcore/Mini_games.lua (hunk 5 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:232]: yep checking nill ill test if your right about being able to remove it.

File: expcore/Mini_games.lua (hunk 7 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:251]: Just forgot to remove that the button was a func at first.

File: expcore/Mini_games.lua (hunk 6 above)
Comment (tovernaar123) [reply expcore/Mini_games.lua:242]: ill use slot_button  cause the work on both versions

**Comment by tovernaar123** [comment]
shood be all good now

File: expcore/Mini_games.lua [hunk 8]
```diff
@@ -183,7 +178,7 @@ function Mini_games.stop_game()
         Event.remove_removable_nth_tick(tick, token)
     end
 
-    for i, player in ipairs(game.connected_players) do
+    for i, _ in ipairs(game.connected_players) do
         game.connected_players[i].teleport({-35,55},"nauvis")
     end
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua:183]: This should be `for _, player in ipairs` and then remove the `game.connected_players[i]`

File: expcore/Mini_games.lua [hunk 9]
```diff
@@ -259,24 +250,22 @@ Gui.element(function(_,parent,name)
...
+    local scroll_table_style = scroll_table.style
     scroll_table_style.top_cell_padding = 3
     scroll_table_style.bottom_cell_padding = 3
-    
-    for i,e in pairs(Mini_games.mini_games) do
-        add_mini_game(scroll_table,i) 
+
+    for i,_ in pairs(Mini_games.mini_games) do
+        add_mini_game(scroll_table,i)
     end
```
Comment (Cooldude2606) [inline expcore/Mini_games.lua]: This can just be `for i in pairs` but really it should be `for name in pairs` since these are the names of mini games.

File: modules/mini-games/Race.lua [hunk 10]
```diff
@@ -73,11 +73,10 @@ Global.register({
...
     gate_boxes = tbl.gate_boxes
     start_players = tbl.start_players
-    
 end)
 
 local function reset_table(table)
-    for i, value in pairs(table) do
+    for i, _ in pairs(table) do
         table[i] = nil
     end
```
Comment (Cooldude2606) [inline modules/mini-games/Race.lua:81]: This can be `for i in pairs` but really it should be `for k in pairs` or `for key in pairs` since this might not be an index because you are using pairs. General rule: i or index is used with ipairs, k or key is used with pairs.

File: modules/mini-games/Race.lua [hunk 11]
```diff
@@ -504,25 +491,28 @@ Gui.element{
...
-    local fuel = dorpdown.get_item(dorpdown.selected_index)
+    local dorpdown_ = flow[dorpdown.name]
+
+    local fuel = dorpdown_.get_item(dorpdown_.selected_index)
     args[1] = fuel
 
-    local laps = flow[text_field_for_laps.name].text
-    args[2] = laps
+    local laps_ = flow[text_field_for_laps.name].text
+    args[2] = laps_
```
Comment (Cooldude2606) [inline modules/mini-games/Race.lua]: Not a blocking change, but people tend to put the _ first, not after. Eg. `_laps`. Although I would suggest using `required_laps` since this will become the number of laps that are required.

File: modules/mini-games/Race.lua (hunk 10 above)
Comment (tovernaar123) [reply modules/mini-games/Race.lua:81]: but then you can argue it shood be i cause i might not be a key but an index (ill change it though)

**Comment by tovernaar123** [comment]
shood be all fixed now
